org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/commons/TextDocument.java

```diff
@@ -139,6 +139,8 @@ public void update(List<TextDocumentContentChangeEvent> changes) {
 			try {
 				synchronized (buffer) {

```

Comment:
Please remove buffer field because the TextDocument stores 2 xml content in this case (buffer and text field). It's not good for memory (2 XML content stored). So the idea is to create StringBuilder instance from the current text in the update method to benefit with StringBuilder#replace  method. and synchronized to a lock field instance (not to the buffer field which must be removed)

...java
private final Object lock = new Object();

...
public void setIncremental(boolean incremental) {
		this.incremental = incremental;
	}

	public boolean isIncremental() {
		return incremental;
	}
...
public void update(List<TextDocumentContentChangeEvent> changes) {
		if (isIncremental()) {
			try {
				StringBuilder buffer = new StringBuilder(getText());
				synchronized (lock) {
...

An another change is to set lineTracker as final because the call of getLineTracker can be created several time when there are several threads which call it in the same time, we could use synchronized but we need not that because lineTracker.set exists as methods.

```java
private final ListLineTracker lineTracker;

	public TextDocument(String text, String uri) {
		this.lineTracker = new ListLineTracker();
		super.setUri(uri);
		setText(text);
	}
       ...

	@Override
	public void setText(String text) {
		super.setText(text);
		lineTracker.set(text);
	}
```

Here the final TextDocument with those 2 changes:

```java
/**
 *  Copyright (c) 2018 Angelo ZERR.
 *  All rights reserved. This program and the accompanying materials
 *  are made available under the terms of the Eclipse Public License v2.0
 *  which accompanies this distribution, and is available at
 *  http://www.eclipse.org/legal/epl-v20.html
 *
 *  Contributors:
 *  Angelo Zerr <angelo.zerr@gmail.com> - initial API and implementation
 */
package org.eclipse.lsp4xml.commons;

import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import org.eclipse.lsp4j.Position;
import org.eclipse.lsp4j.Range;
import org.eclipse.lsp4j.TextDocumentContentChangeEvent;
import org.eclipse.lsp4j.TextDocumentItem;

/**
 * Text document extends LSP4j {@link TextDocumentItem} to provide methods to
 * retrieve position.
 *
 */
public class TextDocument extends TextDocumentItem {

	private static String DEFAULT_DELIMTER = System.lineSeparator();

	private final ListLineTracker lineTracker;

	private boolean incremental;

	private final Object lock = new Object();

	public TextDocument(TextDocumentItem document) {
		this(document.getText(), document.getUri());
		super.setVersion(document.getVersion());
		super.setLanguageId(document.getLanguageId());
	}

	public TextDocument(String text, String uri) {
		this.lineTracker = new ListLineTracker();
		super.setUri(uri);
		setText(text);
	}

	public void setIncremental(boolean incremental) {
		this.incremental = incremental;
	}

	public boolean isIncremental() {
		return incremental;
	}

	@Override
	public void setText(String text) {
		super.setText(text);
		lineTracker.set(text);
	}

	public Position positionAt(int position) throws BadLocationException {
		ListLineTracker lineTracker = getLineTracker();
		return lineTracker.getPositionAt(position);
	}

	public int offsetAt(Position position) throws BadLocationException {
		ListLineTracker lineTracker = getLineTracker();
		return lineTracker.getOffsetAt(position);
	}

	public String lineText(int lineNumber) throws BadLocationException {
		ListLineTracker lineTracker = getLineTracker();
		Line line = lineTracker.getLineInformation(lineNumber);
		String text = super.getText();
		return text.substring(line.offset, line.offset + line.length);
	}

	public String lineDelimiter(int lineNumber) throws BadLocationException {
		ListLineTracker lineTracker = getLineTracker();
		String lineDelimiter = lineTracker.getLineDelimiter(lineNumber);
		if (lineDelimiter == null) {
			if (lineTracker.getNumberOfLines() > 0) {
				lineDelimiter = lineTracker.getLineInformation(0).delimiter;
			}
		}
		if (lineDelimiter == null) {
			lineDelimiter = DEFAULT_DELIMTER;
		}
		return lineDelimiter;
	}

	public Range getWordRangeAt(int textOffset, Pattern wordDefinition) {
		try {
			Position pos = positionAt(textOffset);
			ListLineTracker lineTracker = getLineTracker();
			Line line = lineTracker.getLineInformation(pos.getLine());
			String text = super.getText();
			String lineText = text.substring(line.offset, textOffset);
			int position = lineText.length();
			Matcher m = wordDefinition.matcher(lineText);
			int currentPosition = 0;
			while (currentPosition != position) {
				if (m.find()) {
					currentPosition = m.end();
					if (currentPosition == position) {
						return new Range(new Position(pos.getLine(), m.start()), pos);
					}
				} else {
					currentPosition++;
				}
				m.region(currentPosition, position);
			}
			return new Range(pos, pos);
		} catch (BadLocationException e) {
			return null;
		}
	}

	private ListLineTracker getLineTracker() {
		return lineTracker;
	}

	/**
	 * Update text of the document by using the changes and according the
	 * incremental support.
	 * 
	 * @param changes the text document changes.
	 */
	public void update(List<TextDocumentContentChangeEvent> changes) {
		if (changes.size() < 1) {
			// no changes, ignore it.
			return;
		}
		if (isIncremental()) {
			try {
				StringBuilder buffer = new StringBuilder(getText());
				synchronized (lock) {
					for (TextDocumentContentChangeEvent changeEvent : changes) {
						Range range = changeEvent.getRange();
						int length = 0;

						if (range != null) {
							length = changeEvent.getRangeLength().intValue();
						} else {
							// range is optional and if not given, the whole file content is replaced
							length = buffer.length();
							range = new Range(positionAt(0), positionAt(length));
						}
						String text = changeEvent.getText();
						int startOffset = offsetAt(range.getStart());
						buffer.replace(startOffset, startOffset + length, text);
					}
					setText(buffer.toString());
				}
			} catch (BadLocationException e) {
				// Should never occurs.
			}
		} else {
			// like vscode does, get the last changes
			// see
			// https://github.com/Microsoft/vscode-languageserver-node/blob/master/server/src/main.ts
			TextDocumentContentChangeEvent last = changes.size() > 0 ? changes.get(changes.size() - 1) : null;
			if (last != null) {
				setText(last.getText());
			}
		}
	}

}

```

Reply:
@NikolasKomonen are you sure that you have done the update of TextDocument like I suggest below?

Reply:
@angelozerr I'm getting offset issues with the changes. Did you try running tests with your implementation?

Reply:
?Yes it should work. What is your test (and traces) which failed ?

Reply:
My failures are that it "Couldnt get the position at Offset", and BadLocation errors.

I pushed the changes to this PR, I'm sorry if I missed anything from your changes.